### DISCUSSION TITLE
Switching A/B slots in update.zip

### DISCUSSION THREAD

**Discussion by pascallj** [chenxiaolong/avbroot#249] [Q&A] [answered]
Recently I have been playing with an idea in my head, but I can't seem to figure out if it's theoretically possible or not.

The advantage of having A/B partitioning for the system is that it should be possible to easily test one system and switch back to the old one if it doesn't work. This normally happens automatically. But what if I want to test a system which successfully boots in the other slot and then just want to switch back to the old one. Of course you can sideload the full OTA and it will switch the slot back, but for repeated testing this will take some time and unnecessary wear of the flash.

When the other system has some form of root access, switching slots is possible when booted inside that system. But what if the other system doesn't have root access? When the bootloader is unlocked you can switch the slots in fastboot, but when the bootloader is locked, this is not possible. As far as I'm aware the only way to interact in such a 'low-level' way with the system when the bootloader is locked, is when sideloading a correctly signed update zip.

Do you think it is possible to fabricate some sort of correctly signed update.zip which allows you to switch slots, or perform some kind of shell scripting? Much like how the old updater-script/updater-binary used to work when A/B partitioning wasn't used. Or does this kind of scripting still works on newer devices and we just have to find a way to sign the updates?

**Comment by chenxiaolong** [comment, marked as answer]
> When the other system has some form of root access, switching slots is possible when booted inside that system.

This used to be true, but unfortunately no longer is due to [virtual A/B](https://source.android.com/docs/core/ota/virtual_ab). Now, only partitions that are read by the bootloader, like `boot` and `vendor_boot`, have distinct `_a` and `_b` partitions. The remaining big dm-verity partitions don't.

When flashing an OTA, it takes a snapshot of the current system by setting up some sort of copy-on-write block device for each partition. Any newly written data then gets buffered to files within `/data`. As soon as the new slot boots successfully, the changes are merged and the old "slot" basically ceases to exist.

The only way to boot back into the old slot is to somehow simulate a boot failure to avoid triggering the "this booted successfully" process. This would prevent Android from merging the changes and telling the bootloader that the new slot is good (so that the downgrade protection won't be tripped when switching back).

**Reply by chenxiaolong** [reply]
To answer your questions directly:

> Do you think it is possible to fabricate some sort of correctly signed update.zip which allows you to switch slots

On devices that don't use virtual A/B, I believe this is possible by just creating valid-but-empty `payload.bin`. It would just install nothing and then switch slots like normal.

> or perform some kind of shell scripting?

This is kind of possible. The script needs to live inside one of the partitions in the OTA. Each partition inside `payload.bin` has an optional field that allows an executable to be run: https://github.com/chenxiaolong/avbroot/blob/275b18b2a506c4a1558d6deea3048beb2e02a0df/avbroot/protobuf/update_metadata.proto#L247-L256

The `system` partition, for example, specifies `/system/bin/otapreopt_script`, which precompiles the byte code for every apk to speed up apps' first launch after an update. (But it has a check to only do this when installed from an OTA updater app, not with `adb sideload`)

**Reply by pascallj** [reply]
> > When the other system has some form of root access, switching slots is possible when booted inside that system.
> 
> This used to be true, but unfortunately no longer is due to [virtual A/B](https://source.android.com/docs/core/ota/virtual_ab). Now, only partitions that are read by the bootloader, like `boot` and `vendor_boot`, have distinct `_a` and `_b` partitions. The remaining big dm-verity partitions don't.

Unless I am remembering this wrong, I have used this before on my Pixel 6a succesfully to switch between an avbroot signed rom with Magisk, and one without (I believe by using a fake prepatched, it was before 'rootless' existed). Of course I wasn't able to switch back, but everything else worked fine. But the Pixel 6a does use Virtual A/B, doesn't it?

EDIT: Or does this actually work as expected as the boot.img is the only difference in this case?

**Reply by chenxiaolong** [reply]
> EDIT: Or does this actually work as expected as the boot.img is the only difference in this case?

Yep, I think that's very likely why.

I did a quick hack to print out the `payload.bin` header

```diff
diff --git a/avbroot/src/format/payload.rs b/avbroot/src/format/payload.rs
index c2d3623..ffa1ee4 100644
--- a/avbroot/src/format/payload.rs
+++ b/avbroot/src/format/payload.rs
@@ -164,6 +164,8 @@ impl<R: Read> FromReader<R> for PayloadHeader {
         // Skip manifest signatures.
         reader.read_discard_exact(metadata_signature_size.into())?;
 
+        println!("Manifest: {manifest:#?}");
+
         Ok(Self {
             version,
             manifest,
```

and it looks like the Pixel 6a does indeed use virtual A/B:

```
❯ avbroot ota extract -i bluejay-ota-uq1a.231205.015-2503078c.zip -d stock | grep -A1 vabc
            vabc_enabled: Some(
                true,
--
            vabc_compression_param: Some(
                "gz",
--
            vabc_feature_set: None,
        },
[*] Extracting from the payload: boot, system, vbmeta, vbmeta_system, vbmeta_vendor, vendor_boot
```

**Reply by pascallj** [reply]
But in that case, the whole idea of using A/B as a quick test doesn't really work when virtual partitions are used, unless we want to get involved in creating our own snapshots and stuff like that. That's way too adventurous for my daily driver phone. So I will scrap this idea.

Thank you very much, very useful information :pray:

**Reply by jayanth117** [reply]
@pascallj / @chenxiaolong So can i confirm that, slot switching is not possible with virtual A/B using fastbootd(fastboot <IP ADDRESS> set_active <SLOT>). Only way to switch slots is using OTA, if virtual A/B is enabled?

**Reply by chenxiaolong** [reply]
That's correct.

**Reply by sbpetrack** [reply]
I have spent the last 3-4 weeks convincing my Virtual A/B device to "think" that it is actually a real A/B device. I can even have it stock on "slot A" and Magisk-rooted-stock on "slot B". There are still things I don't really fully understand, and I'm not suggesting that I could have arbitrarily different ROMs in the two slots (at least not yet :)), but (at least for my OnePlus 10 Pro 5G with stock Oxygen 15), I can switch slots when rooted without resorting to fastboot: I can boot to twrp recovery and do it, and when fully booted into system I can use [BootControl](https://github.com/capntrips/BootControl) to switch.  Since I haven't yet got a phone that can do all three tricks at the same time (1. be avbroot-patched; 2. have a locked bootloader; 3. boot :)), I can't assert how this story will end; But unlocked, I can definitely switch.  The essential problem is that there seems to be no way to create the super partition you need via things like lpmake / fastboot, etc. To understand (and some day, produce) the super partition metadata, the [super partition spec in Kaitai Struct](https://formats.kaitai.io/android_super/index.html), together with the [Kaitai WebIDE](https://ide.kaitai.io) were invaluable.  I **truly** don't want to suggest something that might brick your phone, but I found a procedure that uses only "safe, ordinary" tools that (again, on my phone), produces the desired result (again, with a unlocked bootloader -- the following is not ANY kind of "substitute" or "replacement" for avbroot -- it just (maybe) gets you a virtual A/B device where you can "switch slots". You start with a rooted stock ROM, then:
1. use your ordinary system/software updater to install the same ROM 
2. After the install, DON'T REBOOT, instead, just use Magisk to patch the ROM you just installed.
3. Reboot and do exactly the same thing again. 

What essential is that after the successful boot into the "new" ROM, the -cow partitions are NOT erased.
